Complete sixth stage

andrew4herman opened the pull request:
### **CHANGES FROM THE LAST PULL REQUEST**

1. Removed .idea and other unnecessary files from git
2. Added empty line at the end to each file

### **Description**

> Extend your program by adding different algorithms to encode/decode data. The first one would be shifting algorithm (it shifts each letter by the specified number according to its order in the alphabet in circle). The second one would be based on Unicode table, like in the previous stage.
> 
> When starting the program, the necessary algorithm should be specified by an argument (-alg). The first algorithm should be named shift, the second one should be named unicode. If there is no -alg you should default it to shift.
> 
> Remember that in case of shift algorithm you need to encode only English letters (from 'a' to 'z' as the first circle and from 'A' to 'Z' as the second circle i.e. after 'z' comes 'a' and after 'Z' comes 'A').
> 
> To complete this stage, we recommend that you create a set of classes and interfaces for encryption and decryption algorithms.

### **Examples**

**- Example 1**

`java Main -mode enc -in road_to_treasure.txt -out protected.txt -key 5 -alg unicode`

> This command must get data from the file road_to_treasure.txt, encrypt the data with the key 5, create a file called protected.txt and write ciphertext to it.

**- Example 2**

Input:

`java Main -mode enc -key 5 -data "Welcome to hyperskill!" -alg unicode`

Output:

`\jqhtrj%yt%m~ujwxpnqq&`

**- Example 3**

Input:

`java Main -key 5 -alg unicode -data "\jqhtrj%yt%m~ujwxpnqq&" -mode dec`

Output:

`Welcome to hyperskill!`

**- Example 4:**

Input:

`java Main -key 5 -alg shift -data "Welcome to hyperskill!" -mode enc`

Output:

`Bjqhtrj yt mdujwxpnqq!`

**- Example 5:**

Input:

`java Main -key 5 -alg shift -data "Bjqhtrj yt mdujwxpnqq!" -mode dec`

Output:

`Welcome to hyperskill!`

andrew4herman committed:
Complete sixth stage

yvarshavskyigd commented on src/encryptdecrypt/EncryptorDecryptor.java:
```diff
@@ -0,0 +1,98 @@
+            System.out.println(data);
+    }
+
+    private void parseArgs(String[] args) {
```
`EncryptorDecryptor` should not parse arguments

andrew4herman replied:
Fixed

andrew4herman committed:
Add CliParser class

CliParser.java contains useful methods for read and get values from command line arguments.

andrew4herman committed:
Remove responsibility of EncryptorDecryptor to parse arguments

andrew4herman committed:
Resolve warnings

Make necessary fields final,
remove redundant initializers,
unnecessary empty lines.

andrew4herman committed:
Replace responsibility of reading/writing to ReaderWriter class

andrew4herman committed:
Create AbstractCryptor that is base class of all cryptors

andrew4herman committed:
Add close method invocation for writer

andrew4herman committed:
Move CliParser.java to util package

yvarshavskyigd commented on src/encryptdecrypt/Main.java:
```diff
@@ -0,0 +1,27 @@
+                Integer.parseInt(parser.optionValueOrDefault("-key", "0"))
+        );
+
+        cryptor.start();
```
What the contract of this method, `cryptor` can "start" something?

andrew4herman replied:
- [x] Fixed

yvarshavskyigd commented on src/encryptdecrypt/Main.java:
```diff
@@ -0,0 +1,27 @@
+    public static void main(String[] args) {
+        CliParser parser = new CliParser(args);
+
+        AbstractCryptor cryptor = new ConcreteCryptor(
```
What is "ConcreteCryptor"?

andrew4herman replied:
- [X] Removed

yvarshavskyigd commented on src/encryptdecrypt/Main.java:
```diff
@@ -0,0 +1,27 @@
+                parser.optionValueOrDefault("-alg", "shift")
+                        .equals("shift") ? new ShiftAlgorithm() : new UnicodeAlgorithm(),
+
+                parser.optionValueOrDefault("-mode", "enc"),
```
Why cryptor should know what is "enc"?

andrew4herman replied:
- [X] Fixed

yvarshavskyigd commented on src/encryptdecrypt/Main.java:
```diff
@@ -0,0 +1,27 @@
+                        .equals("shift") ? new ShiftAlgorithm() : new UnicodeAlgorithm(),
+
+                parser.optionValueOrDefault("-mode", "enc"),
+                parser.optionValueOrDefault("-data", ""),
```
Does ConcreteCryptor use only once?

andrew4herman replied:
Just like in the task description
- [X] Done

yvarshavskyigd commented on src/encryptdecrypt/Main.java:
```diff
@@ -0,0 +1,27 @@
+
+                parser.optionValueOrDefault("-mode", "enc"),
+                parser.optionValueOrDefault("-data", ""),
+                parser.optionValue("-out"),
```
What is this? Why Cryptor knows about `-out`

andrew4herman replied:
- [X] Fixed

yvarshavskyigd commented on src/encryptdecrypt/Main.java:
```diff
@@ -0,0 +1,27 @@
+                parser.optionValueOrDefault("-mode", "enc"),
+                parser.optionValueOrDefault("-data", ""),
+                parser.optionValue("-out"),
+                parser.optionValue("-in"),
```
The same

andrew4herman replied:
- [X] Fixed

yvarshavskyigd commented on src/encryptdecrypt/cryptors/AbstractCryptor.java:
```diff
@@ -0,0 +1,22 @@
+        }
+    }
+
+    protected abstract void prepareData() throws IOException;
```
What data should cryptor prepare?

andrew4herman replied:
- [X] Improved

yvarshavskyigd commented on src/encryptdecrypt/cryptors/AbstractCryptor.java:
```diff
@@ -0,0 +1,22 @@
+
+    protected abstract void processData();
+
+    protected abstract void outputData() throws IOException;
```
What is output data? again question, does "cryptor" use only once?

andrew4herman replied:
- [X] Fixed

andrew4herman committed:
Add MVC design pattern

andrew4herman committed:
Remove MVC to avoid overengineering

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,41 @@
+            cryptor.setCipher(new UnicodeCipher());
+
+        if (parser.optionOf("-out") != null)
+            FileReaderWriter.writeTo(parser.optionOf("-out"), getCryptedData());
```
What this method do? what crypted data it returns?

andrew4herman replied:
Fixed

yvarshavskyigd replied:
Why you closed this conversation? I don't see your answer

andrew4herman replied:
> What this method do? what crypted data it returns?

Are you talking about getCryptedData() method?

yvarshavskyigd replied:
Yes

andrew4herman replied:
Removed

andrew4herman committed:
Refactor getCryptedData method

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,42 @@
+        if ("unicode".equals(parser.optionOf("-alg")))
+            cryptor.setCipher(new UnicodeCipher());
+
+        String cryptedData = getCryptedOf(readData());
```
Read data from what?

andrew4herman replied:
Fixed

yvarshavskyigd replied:
What does "fixed" mean? I need to know that you understand why I asked it

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Main {
+
+    private static CliParser parser;
```
Is it necessary to keep the state here?

andrew4herman replied:
Removed

yvarshavskyigd replied:
Why removed? maybe you need it?

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,42 @@
+    private static String getCryptedOf(String raw) {
+        int key = Integer.parseInt(parser.optionOrDefault("-key", "0"));
+
+        return switch (parser.optionOf("-mode")) {
```
Do you need here switch?

andrew4herman replied:
Removed

yvarshavskyigd replied:
The same question

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private static String getCryptedOf(String raw) {
+        int key = Integer.parseInt(parser.optionOrDefault("-key", "0"));
```
What happens if parser will be null?

andrew4herman replied:
Should I add "not null" check when I use reference variables?

yvarshavskyigd replied:
Looks like you don't need "reference variables" at all

andrew4herman replied:
Removed this method

yvarshavskyigd replied:
Why you removed this method?

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,42 @@
+        int key = Integer.parseInt(parser.optionOrDefault("-key", "0"));
+
+        return switch (parser.optionOf("-mode")) {
+            case "dec" -> cryptor.decrypt(raw, key);
```
What happens if cryptor will be null?

andrew4herman replied:
Should I add "not null" check when I use cryptor?

yvarshavskyigd replied:
You depend on ctyptor here, which called implicitly

andrew4herman replied:
Fixed

yvarshavskyigd replied:
The same, what "fixed"? let's finish our discussion and then do some changes

andrew4herman committed:
Create outFilePath variable

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,42 @@
+        parser = new CliParser(args);
+        cryptor = new Cryptor();
+
+        if ("unicode".equals(parser.optionOf("-alg")))
```
What happens if statement will be false?

andrew4herman replied:
By the default, Cryptor uses ShiftCipher, instance of which is created inside its constructor

> When starting the program, the necessary algorithm should be specified by an argument (-alg). The first algorithm should be named shift, the second one should be named unicode. If there is no -alg you should default it to shift.

andrew4herman replied:
Fixed

yvarshavskyigd replied:
What fixed?

andrew4herman replied:
> What happens if statement will be false?

The default implementation of Cipher will be used.

> What fixed?

Now we explicitly set the default cipher implementation by passing ShiftCipher instance into Cryptor constructor. So there is no need to create "else" situation

yvarshavskyigd replied:
> The default implementation of Cipher will be used.
How I can know it? What happens if I will pass `baconian` as input parameter?

andrew4herman replied:
Added algorithm selection condition

yvarshavskyigd replied:
I don's see your answer

yvarshavskyigd commented:
DO NOT resolve conversations, only owner can do it!

andrew4herman committed:
Change Cryptor constructor

andrew4herman committed:
Update Main.java

andrew4herman committed:
Update cryptor initialization

yvarshavskyigd commented on src/crypting/CliParser.java:
```diff
@@ -0,0 +1,33 @@
+public class CliParser {
+
+    private final Map<String, String> argValues;
+    private String[] args;
```
Looks like this variable is redundant

andrew4herman replied:
Definitely. Will be removed

andrew4herman replied:
Removed.

yvarshavskyigd commented on src/crypting/CliParser.java:
```diff
@@ -0,0 +1,33 @@
+        parse(arguments);
+    }
+
+    private void parse(String[] arguments) {
```
What happens if the input will be: ?
`-key 5 -alg shift -data "Welcome to hyperskill!" -mode`

andrew4herman replied:
Exception will be thrown. I will add catching block for that

yvarshavskyigd replied:
Do you want to catch unchecked exception? It is not a good idea

andrew4herman replied:
Removed catch block. Now the program shows error and finish in that situation

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,36 @@
+
+    public static void main(String[] args) {
+        CliParser parser = new CliParser(args);
+        Cryptor cryptor;
```
Looks like it should be on line 19

andrew4herman replied:
Agree with you, will replace that

andrew4herman replied:
Done

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,36 @@
+        String data = parser.optionOrDefault("-data", "");
+        int key = Integer.parseInt(parser.optionOrDefault("-key", "0"));
+
+        cryptor = new Cryptor("unicode".equals(algorithm) ?
```
I still have question: what cipher will be using if I pass `Baconian`?

andrew4herman replied:
Task description doesn't provide any information for that, but i suppose that there need to be some exception thrown, right?

yvarshavskyigd replied:
Yes, for example If you pass `Unicode`, but service do not support it and will use `shift` as a default one, would it be good?  For me looks like no 🙂

andrew4herman replied:
You have a point. Added methods for checking every argument and stop program if one of them is invalid.

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,36 @@
+        CliParser parser = new CliParser(args);
+        Cryptor cryptor;
+
+        String outFilePath = parser.optionOf("-out");
```
It would be good to have a class, which will have all these variables and inside will use your cli parser 🙂

andrew4herman replied:
Ok, will encapsulate that in config class

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,36 @@
+                new ShiftCipher()
+        );
+
+        if (data.isEmpty() && inFilePath != null)
```
Complex logic, let's simplify it. 
Correct me if I am wrong, we have both options how to pass data to your service

- Pass by `-data` argument 
- Pass by `-in` argument (data will be inside the file)

So if data is not empty, then we don't care about `-in` argument, right?

Also, data and inFilePath variables are String, but one could be empty and other could be null. Think about it 👌

andrew4herman replied:
But how to act when there will be `-data "" -in "input.txt"?`

yvarshavskyigd replied:
Is it correct way to pass such data to the service? I think then we need to have priority between these arguments or throw exception in that case. What do you think?

andrew4herman replied:
Updated. I have added data validation method and simplified it as we discussed

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,36 @@
+        if (data.isEmpty() && inFilePath != null)
+            data = FileReaderWriter.readFrom(parser.optionOf("-in"));
+
+        String cryptedData = "dec".equals(mode) ?
```
What if I pass mode as `Dec`?

andrew4herman replied:
Should I convert all arguments to lower case (or upper) or throw an exception in that case?
In another words should my program be registry sensible for arguments?

yvarshavskyigd replied:
It depends on requirements which you have in task description 😁
You can do whatever you want 👌

andrew4herman replied:
> You can do whatever you want 👌

Ok, made registry sensible, so now arguments "Dec" and "dec" are different :)

yvarshavskyigd commented:
Much better, good job 👍

I suggest to finish this task on Monday, we can schedule a meeting and do it together. 
What do you think?

andrew4herman commented:
> Much better, good job 👍

Thank you!

> I suggest to finish this task on Monday, we can schedule a meeting and do it together. What do you think?

Sounds good, I will think about your suggestions and prepare my solutions for not to waste your time

andrew4herman committed:
Create Configuration class, add arguments validation check

andrew4herman committed:
Remove toLowerCase arguments converting

andrew4herman committed:
Remove unnecessary throw block

andrew4herman committed:
Make main method decomposition

andrew4herman commented:
@yvarshavskyigd Thank you for your help and suggestions on this project. I have fixed problems you have pointed to. Can you look again please?

yvarshavskyigd commented on src/crypting/Main.java:
```diff
@@ -0,0 +1,40 @@
+
+        try {
+            String inputData = getInputData(config);
+            String cryptedData = getCryptedData(config, cryptor, inputData);
```
Small changes, please rename it to `cryptData`

andrew4herman replied:
Done

andrew4herman committed:
Rename cryptedData to cryptData

andrew4herman committed:
Revert "Rename cryptedData to cryptData"

This reverts commit 24c2293d04ae13128ba2babc152867ba8b2c29d1.

andrew4herman committed:
Rename getCryptedData() to cryptData()

yvarshavskyigd commented:
Approve

yvarshavskyigd commented:
Good job
